velox/docs/develop/spilling.rst

```diff
@@ -0,0 +1,480 @@
+   directly without buffering in the row container;
+5. Store the non-spilled input rows into the row container for hash table build
+   later;
+6. After all the operators finish processing the build inputs, the last
+   finished one will build the hash table from rows collected by all the
+   operators, and send the built table along with the optional spill metadata
+   (if spilling has been triggered) to the hash probe operators through the
+   hash join bridge;
+7. Wait for the spill input to build the next hash table if there is any spill
+   data needed to restore. Otherwise, the hash build operator just finishes.
+   The hash probe operators will pick up a previously spill partition to
+   restore after finishing the join operation;
+8. After receiving the spill input from the hash join bridge, the hash build
+   operator resets the Spiller with advanced partition bits and creates an
+   unordered reader to read build inputs from the spill files set in the spill
+   input;
+9. Go back to step1 to repeat the next hash table build process;
+
+Note that once we have memory arbitration support later, the spilling can also
+be triggered by the memory arbitrator on behalf of any failed memory allocation
+or reservation request from any operators. The memory arbitrator will stop the
+evicted tasks first before spilling so we might not need the coordinator support
+if the spilling is triggered in that case.
+
+HashProbe
+^^^^^^^^^
+
+1. Wait for the next hash table to join from the hash join bridge plus additional spilling metadata if spilling gets involved: the table spill partition id (discussed below)  which identifies the associated spill partition if the table is built from a previously spilled partition and the child spill partition id set which is set to the ids of the partitions that have been spilled while building the hash table;
+2. Create an unordered reader to read probe inputs from the previously spilled data if the table spill partition id is set. The table spill partition id specifies the corresponding spilled probe inputs to read from disk;
+3. Process inputs from either probe input source or the previously spilled probe inputs;
+4. Spill input rows to disk if the corresponding partitions have been spilled by the build side which are identified by child spill partition id set if set;
+5. Join the non-spilled probe input rows with the hash table and produce the output result;
+6. After all the operators finish processing the inputs, if there is no spilled data needed to restore, then all the hash probe operators finish. Otherwise, the last finished operator will send the next spill input to the hash build operators through the hash join bridge;
+7. Go back to step1 to repeat the join process with the next built hash table;
+
+Note that we need to disable some hash probe optimizations if the spilling has
+been triggered by the hash build. For example, we can’t apply dynamic filtering
```

Comment:
> we can’t apply dynamic filtering

dynamic filtering is disabled because the complete set of join keys is not known.